File: ledger-models-python/fintekkers/wrappers/models/security/tips_bond.py
"""TIPSBond wrapper.

Specialises BondSecurity for Treasury Inflation-Protected Securities. The
shared bond fields (coupon, dates, face value) live on
SecurityProto.bond_details; TIPS-specific extras (base CPI, index date,
inflation index type) live on SecurityProto.tips_extension. The two
sub-messages coexist — neither replaces the other.
"""

from datetime import date
from decimal import Decimal
from typing import Optional

from fintekkers.models.security.security_pb2 import (
    SecurityProto,
    BondDetailsProto,
    TipsExtensionProto,
)
from fintekkers.models.security.product_type_pb2 import ProductTypeProto
from fintekkers.models.security.index.index_type_pb2 import IndexTypeProto
from fintekkers.wrappers.models.security.bond_security import (
    BondSecurity,
    _decimal_value,
    _local_date,
)


class TIPSBond(BondSecurity):
    """Wraps a SecurityProto with product_type=TIPS."""

    def get_base_cpi(self) -> Optional[Decimal]:
        self._assert_not_link("base_cpi")
        if not self.proto.HasField("tips_extension"):
            return None
        ext = self.proto.tips_extension
        if not ext.HasField("base_cpi"):
            return None
        value = ext.base_cpi.arbitrary_precision_value
        if value == "":
            return None
        return Decimal(value)

    def get_index_date(self) -> Optional[date]:
        self._assert_not_link("index_date")
        if not self.proto.HasField("tips_extension"):
            return None
        ext = self.proto.tips_extension
        if not ext.HasField("index_date"):
            return None
        d = ext.index_date
        return date(d.year, d.month, d.day)

    def get_inflation_index_type(self) -> IndexTypeProto:
        self._assert_not_link("inflation_index_type")
        if not self.proto.HasField("tips_extension"):
            return IndexTypeProto.UNKNOWN_INDEX_TYPE
        return self.proto.tips_extension.inflation_index_type

    @staticmethod
    def from_pricer_inputs(
        face_value: Decimal,
        coupon_rate: Decimal,
        coupon_type: int,
        coupon_frequency: int,
        issue_date: date,
        maturity_date: date,
        base_cpi: Decimal,
        index_date: date,
        inflation_index_type: int,
    ) -> SecurityProto:
        """Builds a TIPS SecurityProto with both bond_details and
        tips_extension populated, and product_type=TIPS.
        """
        bond_details = BondDetailsProto(
            coupon_rate=_decimal_value(coupon_rate),
            coupon_type=coupon_type,
            coupon_frequency=coupon_frequency,
            face_value=_decimal_value(face_value),
            issue_date=_local_date(issue_date),
            maturity_date=_local_date(maturity_date),
        )
        tips_extension = TipsExtensionProto(
            base_cpi=_decimal_value(base_cpi),
            index_date=_local_date(index_date),
            inflation_index_type=inflation_index_type,
        )
        return SecurityProto(
            product_type=ProductTypeProto.TIPS,
            bond_details=bond_details,
            tips_extension=tips_extension,
        )

File: ledger-models-rust/fintekkers/wrappers/models/tips_bond.rs
//! Typed wrapper around a TIPS-shaped `SecurityProto`.
//!
//! TIPS securities carry shared bond fields on `bond_details` plus
//! TIPS-specific extras (base CPI, index date, inflation index type) on
//! `tips_extension`. This wrapper exposes the TIPS-specific accessors and
//! a builder (`from_pricer_inputs`) that produces a fully-populated
//! `SecurityProto` for the pricer.

use chrono::{Datelike, NaiveDate};
use rust_decimal::Decimal;

use crate::fintekkers::models::security::index::IndexTypeProto;
use crate::fintekkers::models::security::{
    BondDetailsProto, CouponFrequencyProto, CouponTypeProto, ProductTypeProto, SecurityProto,
    TipsExtensionProto,
};
use crate::fintekkers::models::util::{DecimalValueProto, LocalDateProto};
use crate::fintekkers::wrappers::models::utils::errors::Error;
use crate::fintekkers::wrappers::models::utils::serialization::ProtoSerializationUtil;

pub struct TIPSBond {
    pub proto: SecurityProto,
}

impl TIPSBond {
    pub fn from_proto(proto: SecurityProto) -> Result<Self, Error> {
        let pt = ProductTypeProto::from_i32(proto.product_type)
            .unwrap_or(ProductTypeProto::ProductTypeUnknown);
        if pt == ProductTypeProto::Tips {
            Ok(TIPSBond { proto })
        } else {
            Err(Error::NotABondSecurity)
        }
    }

    pub fn base_cpi(&self) -> Option<Decimal> {
        self.proto
            .tips_extension
            .as_ref()
            .and_then(|t| t.base_cpi.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_decimal(d).ok())
    }

    pub fn index_date(&self) -> Option<NaiveDate> {
        self.proto
            .tips_extension
            .as_ref()
            .and_then(|t| t.index_date.as_ref())
            .and_then(|d| ProtoSerializationUtil::deserialize_date(d).ok())
    }

    pub fn inflation_index_type(&self) -> IndexTypeProto {
        self.proto
            .tips_extension
            .as_ref()
            .and_then(|t| IndexTypeProto::from_i32(t.inflation_index_type))
            .unwrap_or(IndexTypeProto::UnknownIndexType)
    }

    pub fn from_pricer_inputs(
        face_value: Decimal,
        coupon_rate: Decimal,
        coupon_type: CouponTypeProto,
        coupon_frequency: CouponFrequencyProto,
        issue_date: NaiveDate,
        maturity_date: NaiveDate,
        base_cpi: Decimal,
        index_date: NaiveDate,
        inflation_index_type: IndexTypeProto,
    ) -> SecurityProto {
        SecurityProto {
            product_type: ProductTypeProto::Tips as i32,
            bond_details: Some(BondDetailsProto {
                coupon_rate: Some(decimal_proto(coupon_rate)),
                coupon_type: coupon_type as i32,
                coupon_frequency: coupon_frequency as i32,
                face_value: Some(decimal_proto(face_value)),
                issue_date: Some(date_proto(issue_date)),
                dated_date: None,
                maturity_date: Some(date_proto(maturity_date)),
                issuance_info: vec![],
            }),
            tips_extension: Some(TipsExtensionProto {
                base_cpi: Some(decimal_proto(base_cpi)),
                index_date: Some(date_proto(index_date)),
                inflation_index_type: inflation_index_type as i32,
            }),
            ..Default::default()
        }
    }
}

fn decimal_proto(d: Decimal) -> DecimalValueProto {
    DecimalValueProto {
        arbitrary_precision_value: d.to_string(),
    }
}

fn date_proto(d: NaiveDate) -> LocalDateProto {
    LocalDateProto {
        year: d.year() as u32,
        month: d.month(),
        day: d.day(),
    }
}
